### PR TITLE
Improve verification for map columns

### PR DESCRIPTION
**Pull request by brendan-driscoll**
Part of #13809

```
== RELEASE NOTES ==

Verifier Changes
* Add checks for keys, values, and cardinality sum when validating a map column.
```

### CODE REVIEW

File: presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/MapColumnValidator.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        }
+
+        if (valueType.isOrderable()) {
+            FunctionCall mapValues = new FunctionCall(QualifiedName.of("map_values"), ImmutableList.of(column.getIdentifier()));
+            FunctionCall sortValues = new FunctionCall(QualifiedName.of("array_sort"), ImmutableList.of(mapValues));
+            checksums.add(new FunctionCall(QualifiedName.of("checksum"), ImmutableList.of(sortValues)));
+        }
+
+        return ImmutableList.of(
+                new SingleColumn(new CoalesceExpression(checksums.build()), Optional.of(delimitedIdentifier(getChecksumColumnAlias(column)))),
```
Comment (mbasmanova) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/MapColumnValidator.java]: This doesn't seem right. I think we need to have multiple columns here, e.g. checksum(m), checksum(array_sort(map_keys(m))) and checksum(array_sort(map_values(m)))

File: presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/MapColumnValidator.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        Object testChecksum = testResult.getChecksum(checksumColumnAlias);
+
+        String cardinalitySumColumnAlias = getCardinalitySumColumnAlias(column);
+        Object controlCardinalitySum = controlResult.getChecksum(cardinalitySumColumnAlias);
+        Object testCardinalitySum = testResult.getChecksum(cardinalitySumColumnAlias);
+
+        return new ColumnMatchResult(
+                Objects.equals(controlChecksum, testChecksum) && Objects.equals(controlCardinalitySum, testCardinalitySum),
+                format(
+                        "control(checksum: %s, cardinality_sum: %s) test(checksum: %s, cardinality_sum: %s)",
```
Comment (mbasmanova) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/MapColumnValidator.java]: Shouldn't there be separate checksums for keys and values?

**Comment by caithagoras** [comment]
Apologize for the delay. I'll get it reviewed by tomorrow morning.

File: presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/MapColumnValidator.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        Object testValuesChecksum = testResult.getChecksum(valuesChecksumColumnAlias);
+
+        String cardinalitySumColumnAlias = getCardinalitySumColumnAlias(column);
+        Object controlCardinalitySum = controlResult.getChecksum(cardinalitySumColumnAlias);
+        Object testCardinalitySum = testResult.getChecksum(cardinalitySumColumnAlias);
+
+        return new ColumnMatchResult(
+                Objects.equals(controlChecksum, testChecksum) && Objects.equals(controlCardinalitySum, testCardinalitySum),
+                format(
+                        "control(checksum: %s, keys_checksum: %s, alues_checksum: %s, cardinality_sum: %s) " +
```
Comment (caithagoras) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/MapColumnValidator.java]: typo: value

File: presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/MapColumnValidator.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+        Expression keysChecksum;
+        FunctionCall mapKeys = new FunctionCall(QualifiedName.of("map_keys"), ImmutableList.of(column.getIdentifier()));
+        if (keyType.isOrderable()) {
+            FunctionCall sortKeys = new FunctionCall(QualifiedName.of("array_sort"), ImmutableList.of(mapKeys));
+            keysChecksum = new FunctionCall(QualifiedName.of("checksum"), ImmutableList.of(sortKeys));
+        }
+        else {
+            keysChecksum = new FunctionCall(QualifiedName.of("checksum"), ImmutableList.of(mapKeys));
+        }
```
Comment (caithagoras) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/MapColumnValidator.java]: Can be simplified as, as below
```
FunctionCall keys = ...
if (keyType.isOrderable()) {
   keys = ...
}
Expression keysChecksum = new FunctionCall(QualifiedName.of("checksum"), ImmutableList.of(keys));

File: presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/MapColumnValidator.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        String valuesChecksumColumnAlias = getValuesChecksumColumnAlias(column);
+        Object controlValuesChecksum = controlResult.getChecksum(valuesChecksumColumnAlias);
+        Object testValuesChecksum = testResult.getChecksum(valuesChecksumColumnAlias);
+
+        String cardinalitySumColumnAlias = getCardinalitySumColumnAlias(column);
+        Object controlCardinalitySum = controlResult.getChecksum(cardinalitySumColumnAlias);
+        Object testCardinalitySum = testResult.getChecksum(cardinalitySumColumnAlias);
+
+        return new ColumnMatchResult(
+                Objects.equals(controlChecksum, testChecksum) && Objects.equals(controlCardinalitySum, testCardinalitySum),
```
Comment (caithagoras) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/MapColumnValidator.java]: Also make sure key checksum matches and and value checksum matches.
nit: one condition per line for long line:
```Java
Object.equals(...)
    && Object.equals(...)
    && Object.equals(...)
    && Object.equals(...)
```

**Comment by caithagoras** [comment]
Potentially, if key/values types are of other non-simple types, we'll be able to apply the respective column validator on the keys and values (with some restrictions).
Depends on https://github.com/prestodb/presto/pull/14036

**Comment by caithagoras** [comment]
@brendan-driscoll Please squash the 2 commits.

File: presto-verifier/src/test/java/com/facebook/presto/verifier/checksum/TestChecksumValidator.java [hunk 6]
```diff
@@ -297,9 +304,9 @@ public void testArray()
         ChecksumResult controlChecksum = new ChecksumResult(
                 5,
                 ImmutableMap.<String, Object>builder()
-                        .put("int_array_checksum", new SqlVarbinary(new byte[] {0xa}))
+                        .put("int_array_checksum", new SqlVarbinary(new byte[]{0xa}))
```
Comment (caithagoras) [inline presto-verifier/src/test/java/com/facebook/presto/verifier/checksum/TestChecksumValidator.java]: Your formatting is off. Please follow 7e and 7f  in https://our.internmc.facebook.com/intern/wiki/DataInfra/Presto/Source/
to install Airlift code style, and then format the file with (option + cmd + L).

File: presto-verifier/src/test/java/com/facebook/presto/verifier/checksum/TestChecksumValidator.java [hunk 7]
```diff
@@ -387,4 +394,70 @@ public void testRow()
...
+                        .put("map_checksum", new SqlVarbinary(new byte[]{0x1a}))
+                        .put("map_keys_checksum", new SqlVarbinary(new byte[]{0x0b}))
+                        .put("map_values_checksum", new SqlVarbinary(new byte[]{0x0c}))
+                        .put("map_cardinality_sum", 3L)
+                        .build());
+        assertEquals(
+                checksumValidator.getMismatchedColumns(columns, controlChecksum, testChecksum),
+                ImmutableMap.builder()
+                        .put(MAP_COLUMN, new ColumnMatchResult(false, "control(checksum: 0a, keys_checksum: 0b, values_checksum: 0c, cardinality_sum: 3) " +
+                                "test(checksum: 1a, keys_checksum: 0b, values_checksum: 0c, cardinality_sum: 3)"))
```
Comment (caithagoras) [inline presto-verifier/src/test/java/com/facebook/presto/verifier/checksum/TestChecksumValidator.java]: nit: all arguments on the same line, or one argument per line. Same below.

**Comment by caithagoras** [comment]
Looks good % nit.
As mentioned earlier, please squash the 3 commits into 1 using interactive rebase: https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history

Commit tile: "Add additional checks when validating map columns".

Add a release notes section similar to https://github.com/prestodb/presto/pull/13916#issue-358780198

" Add checks for key, values, and cardinalities sum when validating a map column."

**Comment by brendan-driscoll** [comment]
`== RELEASE NOTES ==

Verifier Changes
* Add checks for keys, values, and cardinality sum when validating a map column.`

**Comment by mbasmanova** [comment]
@brendan-driscoll Would you please squash commits?

**Comment by viczhang861** [comment]
@brendan-driscoll Why is this a fix for #3809 ?

**Comment by mbasmanova** [comment]
@brendan-driscoll Thanks for squashing commits. Please, update the commit to match the guidelines at https://chris.beams.io/posts/git-commit/

For example,

```
Improve verification for map columns

Add checksums for map key, values and map sizes.
```

**Comment by caithagoras** [comment]
Please update the commit title and the PR title.

**Comment by brendan-driscoll** [comment]
> @brendan-driscoll Why is this a fix for #3809 ?

That was a typo, should have been #13809

**Comment by mbasmanova** [comment]
@brendan-driscoll There are some merge conflicts. Would you rebase on top of latest master and update the PR?

**Comment by mbasmanova** [comment]
@brendan-driscoll Test failures seem to be related to the changes. Would you take a look?

File: presto-verifier/src/test/java/com/facebook/presto/verifier/checksum/TestChecksumValidator.java [hunk 8]
```diff
@@ -387,6 +393,78 @@ public void testRow()
...
+                        .build());
+
+        expectedOutput = "control(checksum: 0a, keys_checksum: 0b, values_checksum: 0c, cardinality_sum: 3) " +
+                "test(checksum: 0a, keys_checksum: 0b, values_checksum: 0c, cardinality_sum: 4)";
+        assertEquals(
+                checksumValidator.getMismatchedColumns(columns, controlChecksum, testChecksum),
+                ImmutableMap.builder()
+                        .put(MAP_COLUMN, new ColumnMatchResult(false, MAP_COLUMN, expectedOutput))
+                        .build());
+    }
```
Comment (caithagoras) [inline presto-verifier/src/test/java/com/facebook/presto/verifier/checksum/TestChecksumValidator.java]: You can combine those mismatched cases into one.
Also, add a test case where the key/value are not orderable and hence the `array_sort` function call is not used.

File: presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/MapColumnValidator.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.String.format;
+
+public class MapColumnValidator
+        implements ColumnValidator
+{
+    @Inject
+    public MapColumnValidator()
+    {
+    }
```
Comment (caithagoras) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/MapColumnValidator.java]: Usually, one line for empty constructor:
```Java
public MapColumnValidator() {}
```

**Pull request merged**